### DISCUSSION TITLE
Make FromKitShema genric

### DISCUSSION THREAD

**Discussion by ppierre** [formkit/formkit#741] [Ideas]
```ts
type FormKitS<T> = FormKitSchemaFormKit &
  (
    | {
        name: keyof T
      }
    | {
        ignore: true
      }
  )
```

The idea is simply strongly typing the schema and the name of the inputs field.

I didn't refine it more as I'm not familiar with FormKit schemas and FormKit file field is a bit of pain.

**Comment by justin-schroeder** [comment]
FormKitSchemas are already pretty strongly typed — I think I would need more detail on what artifact you’re trying to extract exactly?

**Comment by ppierre** [comment]
I try to do the same as `definesProps<MyType>()`in Vue 3.3 
```ts
interface HouseRecord {
  houseName:string,
  surface:number,
  /*...*/
}

type FormKitS<T> = FormKitSchemaNode &
{
  name: keyof T
  $formkit: string
}

const schema: FormKitS<HouseRecord>[] = [
  {
    name: 'houseName',
    $formkit: 'text',
    label: 'name of the house'
  },
  {
    name: 'surface',
    $formkit: 'number',
    label: 'Surface'
  },
/*...*/
]
```

By doing this the `name` value are checked against my data type (generated from the SQL columns).

**Reply by justin-schroeder** [reply]
I could see this being useful in some circumstances yes. Although this type is sort of a "lie" since it depends on runtime details. It does ensure that the schema would not be initially composed with any wrong names though, so I suppose that would be beneficial at least to some degree. Is `name` the only thing that would be typed?

**Reply by ppierre** [reply]
It may be possible to restrict `$formkit` value depending on the type of the key ? I don't know if typescript can do this. But it would be more restrictive than useful IMHO.

Typing `name` is really useful BC it helps auto-completion a lot ! (And sorry I didn't intent to make a mistake in my example (`name` / `houseName`).
